plugins/outputs/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,253 @@
+  enable_sniffer = true
+  ## Set the interval to check if the nodes are available, in seconds
+  ## Setting to 0 will disable the health check (not recommended in production)
+  health_check_interval = 10
+  ## HTTP basic authentication details (eg. when using Shield)
+  # username = "telegraf"
+  # password = "mypassword"
+
+  # Index Config
+  ## The target index for metrics (Elasticsearch will create if it not exists).
+  ## You can use the date specifiers below to create indexes per time frame.
+  ## The metric timestamp will be used to decide the destination index name
+  # %Y - year (2016)
+  # %y - last two digits of year (00..99)
+  # %m - month (01..12)
+  # %d - day of month (e.g., 01)
+  # %H - hour (00..23)
+  index_name = "telegraf-%Y.%m.%d" # required.
+
+  ## Template Config
+  ## Set to true if you want telegraf to manage its index template.
+  ## If enabled it will create a recommended index template for telegraf indexes
+  manage_template = true
+  ## The template name used for telegraf indexes
+  template_name = "telegraf"
+  ## Set to true if you want to overwrite an existing template
+  overwrite_template = false
+`
+
+func (a *Elasticsearch) Connect() error {
+	if a.URLs == nil || a.IndexName == "" {
+		return fmt.Errorf("Elasticsearch urls or index_name is not defined")
+	}
+
+	ctx := context.Background()
+
+	var clientOptions []elastic.ClientOptionFunc
+
+	clientOptions = append(clientOptions,
+		elastic.SetSniff(a.EnableSniffer),
+		elastic.SetURL(a.URLs...),
+	)
+
+	if a.Username != "" && a.Password != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetBasicAuth(a.Username, a.Password),
+		)
+	}
+
+	if a.HealthCheckInterval > 0 {
+		clientOptions = append(clientOptions,
+			elastic.SetHealthcheckInterval(time.Duration(a.HealthCheckInterval)*time.Second),
+		)
+	}
+
+	client, err := elastic.NewClient(clientOptions...)
+
+	if err != nil {
+		return fmt.Errorf("Elasticsearch connection failed: %s", err)
+	}
+
+	// check for version on first node
+	esVersion, err := client.ElasticsearchVersion(a.URLs[0])
+
+	if err != nil {
+		return fmt.Errorf("Elasticsearch version check failed: %s", err)
+	}
+
+	// warn about ES version
+	if i, err := strconv.Atoi(strings.Split(esVersion, ".")[0]); err == nil {
+		if i < 5 {
+			log.Println("W! Elasticsearch version not supported: " + esVersion)
+		} else {
+			log.Println("I! Elasticsearch version: " + esVersion)
+		}
+	}
+
+	a.Client = client
+
+	if a.ManageTemplate {
+		if a.TemplateName == "" {
+			return fmt.Errorf("Elasticsearch template_name configuration not defined")
+		}
+
+		templateExists, errExists := a.Client.IndexTemplateExists(a.TemplateName).Do(ctx)
+
+		if errExists != nil {
+			return fmt.Errorf("Elasticsearch template check failed, template name: %s, error: %s", a.TemplateName, errExists)
+		}
+
+		if (a.OverwriteTemplate) || (!templateExists) {
+			// Create or update the template
+			tmpl := fmt.Sprintf(`
+			{ "template":"%s*",
```

Comment:
The template is kind of optional, the plugin will work without it and ES would detect correctly most of the types, and even if not optimal, it could still work for querying and graphing the data. 

One of the drawbacks to not use this template is to have the tags analyzed by ES. In this case, using the default mapping from ES, instead of using `tag.host` one would have to use `tag.host.keyword` to access the non-analyzed field.

The idea is that the template provided will have sane defaults/types. Everything can be override by another template if the ES admin wants, or have this template updated manually. 

Maybe I can add in the doc that it is possible to check the template created by issuing `curl http://localhost:9200/_template/telegraf?pretty` (considering the template name is telegraf)

BTW I need to check few additional things about the mapping of the number fields. I saw today some fields being mapped as float and others as long, and I need to check if this is correct and/or it causes any trouble.